conbench/entities/benchmark_result.py

```diff
@@ -223,38 +226,27 @@ def create(userres) -> "BenchmarkResult":
         case = Case.get_or_create({"name": benchmark_name, "tags": tags})
         context = Context.get_or_create({"tags": userres["context"]})
         info = Info.get_or_create({"tags": userres.get("info", {})})
-
-        # Create a corresponding `Run` entity in the database if it doesn't
-        # exist yet. Use the user-given `id` (string) as primary key. If the
-        # Run is already known in the database then only update the
-        # `has_errors` property, if necessary. All other run-specific
-        # properties provided as part of this BenchmarkResultCreate structure
-        # (like `machine_info` and `run_name`) get silently ignored.
-        run = Run.first(id=userres["run_id"])
-        if run:
-            if "error" in userres:
-                run.has_errors = True
-                run.save()
+        if "machine_info" in userres:
+            hardware = Machine.get_or_create(userres["machine_info"])
         else:
-            hardware_info_field = (
-                "machine_info" if "machine_info" in userres else "cluster_info"
-            )
-            Run.create(
-                {
-                    "id": userres["run_id"],
-                    "name": userres.pop("run_name", None),
-                    "reason": userres.pop("run_reason", None),
-                    "github": userres.pop("github", None),
-                    hardware_info_field: userres.pop(hardware_info_field),
-                    "has_errors": "error" in userres,
-                }
+            hardware = Cluster.get_or_create(userres["cluster_info"])
+
+        user_given_commit_info: Optional[TypeCommitInfoGitHub] = userres.get("github")
+        commit = None
+        if user_given_commit_info is not None:
+            commit = commit_fetch_info_and_create_in_db_if_not_exists(
+                user_given_commit_info
             )
-            # The above's `create()` might fail (race condition), in which case
-            # we can re-read, but then we also have to call
-            # `validate_run_result_consistency(userres)` one more time (because
-            # _we_ are not the ones who created the Run).
 
         result_data_for_db["run_id"] = userres["run_id"]
+        result_data_for_db["run_tags"] = userres.get("run_tags") or {}
```

Comment:
Do we have a thought process on `None` (nullable) vs `{}`?

Reply:
I think I'd rather it always be stored in the database as `{}` so it's a little easier to work with in code because it's always a `dict`. So I made the column non-nullable. 

However, IMO it shouldn't be required to POST `"run_tags": {}` if you don't have any run tags. So this line turns whatever was posted into `{}` if it was missing (or otherwise falsey). Does this reasoning make sense to you?

Reply:
> IMO it shouldn't be required to POST "run_tags": {} if you don't have any run tags

Fully agree.

I don't know the answer really here yet; I just think it might be an important decision to make.

What might matter:

- In the DB, a JSON `null` value or a database `NULL` value (not the same) might make certain queries potentially more efficient than storing an empty JSON object (this is speculation).
- Easy-to-reason about (we agree on this goal). Example, I think when working in code I could like the following simplicity: when it's a dictionary then it it's a flat string/string mapping with at least one key (I would love for us to add this constraint if we can, to work against the nesting). And then there should be a precise, explicit way to check for "not set"; and here I often like the `is None` identity check in Python.

In what I wrote above I think the question of strictness in the JSON document might be more important to figure out initially. _Can_ we do a flat str/str mapping?

Reply:
Great, I'll add some validation then!

>like `timestamp`

@alistaire47 reminded me that clients couldn't ever set `run.timestamp` anyway, so really the only migration people might have to do is for the `info` field. (We're dropping the notion of the other mutable fields in this PR; see my upcoming comment)